crates/subspace-networking/src/behavior/persistent_parameters.rs

```diff
@@ -0,0 +1,248 @@
+pub struct NetworkingParametersManager {
+    // LRU cache for the known peers and their addresses
+    known_peers: LruCache<PeerId, HashSet<Multiaddr>>,
+    // Period between networking parameters saves.
+    networking_parameters_save_delay: Pin<Box<Fuse<Sleep>>>,
+    // Parity DB instance
+    db: Arc<Db>,
+    // Column ID to persist parameters
+    column_id: u8,
+    // Key to persistent parameters
+    object_id: &'static [u8],
+}
+
+impl NetworkingParametersManager {
+    /// Object constructor. It accepts `NetworkingParametersProvider` implementation as a parameter.
+    /// On object creation it starts a job for networking parameters cache handling.
+    pub fn new(path: &Path) -> Result<Self, NetworkParametersPersistenceError> {
+        let mut options = Options::with_columns(path, 1);
+        // We don't use stats
+        options.stats = false;
+
+        let db = Db::open_or_create(&options)?;
+        let column_id = 0u8;
+        let object_id = b"global_networking_parameters_key";
+
+        // load known peeers cache.
+        let cache = db
+            .get(column_id, object_id)?
+            .map(|data| {
+                let result = serde_json::from_slice::<NetworkingParametersDto>(&data).map(|data| {
+                    let cache: LruCache<PeerId, HashSet<Multiaddr>> = data.into();
+
+                    cache
+                });
+
+                if result.is_ok() {
+                    trace!("Networking parameters loaded from DB");
+                }
+
+                result
+            })
+            .unwrap_or_else(|| Ok(LruCache::new(PEER_CACHE_SIZE)))?;
+
+        Ok(Self {
+            db: Arc::new(db),
+            column_id,
+            object_id,
+            known_peers: cache,
+            networking_parameters_save_delay: Self::default_delay(),
+        })
+    }
+
+    // Helps create a copy of the internal LruCache
+    fn clone_known_peers(&self) -> LruCache<PeerId, HashSet<Multiaddr>> {
+        let mut known_peers = LruCache::new(self.known_peers.cap());
+
+        for (peer_id, addresses) in self.known_peers.iter() {
+            known_peers.push(*peer_id, addresses.clone());
+        }
+
+        known_peers
+    }
+
+    /// Creates a reference to the `NetworkingParametersRegistry` trait implementation.
+    pub fn boxed(self) -> Box<dyn NetworkingParametersRegistry> {
+        Box::new(self)
+    }
+
+    // Create default delay for networking parameters.
+    fn default_delay() -> Pin<Box<Fuse<Sleep>>> {
+        Box::pin(sleep(Duration::from_secs(DATA_FLUSH_DURATION_SECS)).fuse())
+    }
+}
+
+#[async_trait]
+impl NetworkingParametersRegistry for NetworkingParametersManager {
+    async fn add_known_peer(&mut self, peer_id: PeerId, addresses: Vec<Multiaddr>) {
+        let addr_set = addresses.iter().cloned().collect::<HashSet<_>>();
+
+        if let Some(addresses) = self.known_peers.get_mut(&peer_id) {
+            *addresses = addresses.union(&addr_set).cloned().collect()
```

Comment:
I think here we also want to have LRU, otherwise addresses of pathological peers can flood these hashsets.

Reply:
Makes sense. I'll add it.